splice_machine/src/main/java/com/splicemachine/derby/impl/sql/catalog/upgrade/SpliceCatalogUpgradeScripts.java

```diff
@@ -33,56 +30,81 @@
     protected static final Logger LOG=Logger.getLogger(SpliceCatalogUpgradeScripts.class);
 
     SpliceDataDictionary sdd;
-    Splice_DD_Version catalogVersion;
     TransactionController tc;
-    TreeMap<Splice_DD_Version, UpgradeScript> scripts;
-    Comparator<Splice_DD_Version> ddComparator;
-    
-    public SpliceCatalogUpgradeScripts(SpliceDataDictionary sdd,Splice_DD_Version catalogVersion,TransactionController tc){
+
+    List<VersionAndUpgrade> scripts;
```

Comment:
Instead of using a `List` and perform a linear search each time we want to perform an upgrade, I would keep the tree structure used before and replace it with something like Guava's [`MultiMap`](https://guava.dev/releases/19.0/api/docs/com/google/common/collect/Multimap.html) instead.

Reply:
i made this especially easy since the whole upgrade path is really hard to test, people modify it, and if it's broken it's a high cost, while optimal performance is lower priority here. last version was using a sophisticated map, then someone changed it to still iterate over all element.
this upgrade is called ONCE when database is started, the number of items in the list is currently 22. Even for 100 items i would doubt that creating a multimap + searching is faster than just looping over a list. even if - I would really like to keep this rock-solid-bullet-proof code, and not trying to squeeze out a second or two.

Reply:
I agree with @martinrupp , I think it's ok to have a list here since this code is not performance sensitive (at least for now)

Reply:
> I would really like to keep this rock-solid-bullet-proof code, and not trying to squeeze out a second or two.

I think these are mutually exclusive.

Anyway, it is a fair point that the number of items is too small that the overhead is negligible.

Reply:
> > I would really like to keep this rock-solid-bullet-proof code, and not trying to squeeze out a second or two.

> I think these are not mutually exclusive

yes, i agree.